_posts/2021-09-07-what-javascript-tests-could-learn-from-rspec.md

```diff
@@ -0,0 +1,278 @@
+tested...and `context` to describe scenarios of the thing being tested.
+
+For example, if I wanted to test the `multiply` method of a `Calculator` class, I might write some test scenarios
+that look like this:
+
+```rb
+describe "Calculator" do
+  describe ".multiply" do
+    context "when the first value is negative" do
+      context "when the second value is negative" do
+        it "returns a positive number" do
+        end
+      end
+      context "when the second value is positive" do
+        it "returns a negative number" do
+        end
+      end
+    end
+  end
+end
+```
+
+See the difference in those test cases between `describe` and `context`? The way I think about it is: if the
+statement coming after my `describe`/`context` describes a pre-condition for the test, it's a `context`; otherwise
+it's a `describe`.
+
+`context` wouldn't be hard to implement in JavaScript — I'd bet there are test frameworks that have it. It'd just
+be an alias of `describe`.
+
+## 2. `let` blocks
+
+[`let` blocks](https://relishapp.com/rspec/rspec-core/v/2-11/docs/helper-methods/let-and-let) are used in an RSpec
+test to set things up for your test scenario.
+
+Here's a test for a `Counter` class, verifying that when I call the `increment` method on an instance, its stored
+value becomes `1`.
+
+```rb
+describe "counter" do
+
+  let(:counter) { Counter.new }
+
+  describe "increment" do
+    it "increments by 1" do
+      counter.increment
+
+      counter.value.should eq(1)
+    end
+  end
+end
+```
+
+If you're new to Ruby, the only line that doesn't translate almost directly to a similar JavaScript expression is
+the `let` statement.
+
+The `let` statement in RSpec [creates a method with a specified name, which lazily evaluates to the result of a
+block][rspec-let]. In this case, we get a method named `counter`, which is evaluated to a new instance of the
+`Counter` class. There are a few important things to note about `let` blocks:
+
+1. They're evaluated lazily (by default). That `counter` doesn't actually get created until I reference it.
+2. They're memoized. Wherever I reference `counter` within that `describe "counter"` block, I'm getting the same
+   instance. It's initialized to whatever I return inside the `let` block.
+3. I can override a `let` block deeper inside the tree of tests, by declaring another `let(:counter)` later. When I
+   do this, the closest `let` block in the tree for that thing is the one that gets used.
+
+I don't think it's possible to implement `let` in JavaScript — at least not in the way it exists in RSpec. It
+relies on
+[Ruby meta-programming to intercept calls to missing methods](https://www.leighhalliday.com/ruby-metaprogramming-method-missing),
+which just doesn't exist in JavaScript. The [givens](https://github.com/enova/givens) library does something pretty
+close, but it relies on string keys to define things, and there's a bit of extra work when working with TypeScript.
```

Comment:
I think one could get pretty close to this via a clever use of [Proxies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy)